// @gravity-ui/components/build/esm/components/DefinitionList/DefinitionList.js
import React from 'react';
import { ClipboardButton } from '@gravity-ui/uikit';
import { HelpPopover } from '../HelpPopover';
import { block } from '../utils/cn';
import { isUnbreakableOver } from './utils';
import './DefinitionList.css';
export const b = block('definition-list');
function getTitle(title, content) {
    if (title) {
        return title;
    }
    if (typeof content === 'string' || typeof content === 'number') {
        return String(content);
    }
    return undefined;
}
function getNoteElement(note) {
    let noteElement = null;
    const popoverClassName = b('item-note-tooltip');
    if (note) {
        if (typeof note === 'string') {
            noteElement = (React.createElement(HelpPopover, { className: popoverClassName, content: note, placement: ['bottom', 'top'] }));
        }
        if (typeof note === 'object') {
            noteElement = (React.createElement(HelpPopover, Object.assign({ className: popoverClassName, placement: ['bottom', 'top'] }, note)));
        }
    }
    return noteElement;
}
export function DefinitionList({ items, responsive, nameMaxWidth, contentMaxWidth = 'auto', className, itemClassName, copyPosition = 'outside', qa, }) {
    const keyStyle = nameMaxWidth
        ? {
            flexBasis: nameMaxWidth,
        }
        : {};
    const valueStyle = typeof contentMaxWidth === 'number'
        ? {
            flexBasis: contentMaxWidth,
            maxWidth: contentMaxWidth,
        }
        : {};
    const normalizedItems = React.useMemo(() => {
        return items.map((value, index) => (Object.assign(Object.assign({}, value), { key: index })));
    }, [items]);
    return (React.createElement("dl", { className: b({ responsive }, className), "data-qa": qa }, normalizedItems.map(({ name, key, content, contentTitle, nameTitle, copyText, note, multilineName }) => {
        const definitionContent = content !== null && content !== void 0 ? content : '—';
        const iconInside = copyPosition === 'inside';
        const definition = copyText ? (React.createElement("div", { className: b('copy-container', { 'icon-inside': iconInside }) },
            React.createElement("span", null, definitionContent),
            React.createElement(ClipboardButton, { size: "s", text: copyText, className: b('copy-button'), view: iconInside ? 'raised' : 'flat-secondary' }))) : (definitionContent);
        const noteElement = (React.createElement(React.Fragment, null,
            "\u00A0",
            getNoteElement(note)));
        return (React.createElement("div", { key: key, className: b('item', itemClassName) },
            React.createElement("dt", { className: b('term-container', { multiline: multilineName }), style: keyStyle },
                React.createElement("div", { className: b('term-wrapper') },
                    React.createElement("span", { title: getTitle(nameTitle, name) }, name),
                    multilineName && noteElement),
                !multilineName && noteElement,
                React.createElement("div", { className: b('dots', { 'with-note': Boolean(note) }) })),
            React.createElement("dd", { className: b('definition'), title: getTitle(contentTitle, content), style: Object.assign(Object.assign({}, valueStyle), { lineBreak: typeof content === 'string' &&
                        isUnbreakableOver(20)(content)
                        ? 'anywhere'
                        : undefined }) }, definition)));
    })));
}
